bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/checksum/DigestManager.java

```diff
@@ -89,17 +107,27 @@ public static DigestManager instantiate(long ledgerId, byte[] passwd, DigestType
      */
     public ByteBufList computeDigestAndPackageForSending(long entryId, long lastAddConfirmed, long length,
             ByteBuf data) {
-        ByteBuf headersBuffer = PooledByteBufAllocator.DEFAULT.buffer(METADATA_LENGTH + macCodeLength);
-        headersBuffer.writeLong(ledgerId);
-        headersBuffer.writeLong(entryId);
-        headersBuffer.writeLong(lastAddConfirmed);
-        headersBuffer.writeLong(length);
+        ByteBuf sendBuffer = byteBufAllocator.buffer(METADATA_LENGTH + macCodeLength + data.readableBytes());
+        sendBuffer.writeLong(ledgerId);
+        sendBuffer.writeLong(entryId);
+        sendBuffer.writeLong(lastAddConfirmed);
+        sendBuffer.writeLong(length);
 
-        update(headersBuffer);
+        update(sendBuffer);
         update(data);
-        populateValueAndReset(headersBuffer);
+        populateValueAndReset(sendBuffer);
+
+        if (data.hasArray()) {
+            // fast path: single copy into existing array
+            sendBuffer.writeBytes(data.array(), data.arrayOffset(), data.readableBytes());
+        } else {
+            // fall-back to slow path with extra object allocation and copy
+            byte[] sendArray = new byte[sendBuffer.readableBytes()];
+            sendBuffer.getBytes(sendBuffer.readerIndex(), sendArray);
+            sendBuffer.writeBytes(sendArray);
+        }
 
-        return ByteBufList.get(headersBuffer, data);
+        return ByteBufList.get(sendBuffer);
```

Comment:
Maybe in a follow up change we can get rid of ByteBufList and use directly the ByteBuf.

Cc @merlimat

Reply:
@eolivelli no. see my comment.

Reply:
@eolivelli this would make an additional copy of the payload buffer